File: 백준/Gold/2230. 수 고르기/수 고르기.py
#차이가 m이상이면서 가장 작은거 -> 차이를 기준으로 투 포인터 활용
import sys
n, m = map(int, sys.stdin.readline().rstrip().split())

a = []
for _ in range(n):
  a.append(int(sys.stdin.readline().rstrip()))

a.sort()
start, end = 0, 1

current_gap = float('inf')
while start < n and end < n:
  if a[end] - a[start] >= m:
    current_gap = min(current_gap, a[end] - a[start])
    start += 1
    end = start
  elif a[end] - a[start] < m:
    end += 1
  
print(current_gap)

File: 백준/Gold/2230. 수 고르기/수 고르기.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeSet;


public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] input = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]);
        int m = Integer.parseInt(input[1]);
        int[] arr = new int[n];
        for(int i=-0; i<n;i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        Arrays.sort(arr);
        int start = 0;
        int end = 0;
        int result = Integer.MAX_VALUE;
        while(end < n && start <= end) {
            int gap = arr[end] - arr[start];
            if(gap < m) {
                end += 1;
            } else if(gap >= m) {
                start += 1;
                result = Math.min(result, gap);
            }
        }
        System.out.println(result);

    }
}

